feat(ops): add AI Contract Test Maintenance System v0.1.0

aliersh opened the pull request:
# Add AI Contract Test Maintenance System v0.1.0

## Overview

This PR introduces the **AI Contract Test Maintenance System**, an automated toolchain designed to continuously reduce technical debt in contract tests within the `contracts-bedrock` monorepo. The system integrates with an AI coding agent to generate pull requests that improve test quality and coverage.

## What's in this PR (v0.1.0)

### Test Ranking System
- **Test selection** based on staleness metrics between test files and their corresponding source contracts
- **Two-branch scoring algorithm** that prioritizes tests whose contracts have diverged ahead, then falls back to test age
- **Git-based analysis** using commit timestamps to calculate staleness and priority scores.

## Sample Output

The system generates a ranked JSON file with entries like:

```json
{
  "generated_at": "2025-09-22T...",
  "entries": [
    {
      "test_path": "test/L1/ProtocolVersions.t.sol",
      "contract_path": "src/L1/ProtocolVersions.sol", 
      "staleness_days": -98.21,
      "score": 135.84
    }
  ]
}
```

Higher scores indicate higher priority for maintenance attention.

## Documentation

- **Operational Runbook** (`docs/runbook.md`) - System overview, usage, and troubleshooting

aliersh committed:
feat(ops): add AI Contract Test Maintenance System v0.1.0

- Add new ai-eng directory for AI engineering projects
- Implement test ranking system with staleness-based scoring algorithm
- Add modular Python architecture for test file discovery and contract mapping
- Integrate with existing exclusion system from contracts-bedrock
- Add runbook.md document
- Add justfile integration for command-line execution
- Generate JSON output format for ranked test results
- Support git-based commit timestamp analysis for priority calculation
- Add CHANGELOG.md following

smartcontracts commented on ops/ai-eng/contracts-test-maintenance/tests_ranker/output/ranking.json:
```diff
@@ -0,0 +1,557 @@
+{
```
We probably don't need to commit this - CI should be able to run the script and use it at runtime rather than having it committed

aliersh replied:
Fixed in [5d4ab0f](https://github.com/ethereum-optimism/optimism/pull/17560/commits/5d4ab0f3f51cbb4704b7f60ca2647b86c94379f7)

smartcontracts commented on ops/ai-eng/contracts-test-maintenance/CHANGELOG.md:
```diff
@@ -0,0 +1,34 @@
+# Changelog
```
Changelog might be annoying to maintain, probably easier to not have it honestly

aliersh replied:
Fixed in [5d4ab0f](https://github.com/ethereum-optimism/optimism/pull/17560/commits/5d4ab0f3f51cbb4704b7f60ca2647b86c94379f7)

smartcontracts commented on ops/ai-eng/contracts-test-maintenance/docs/runbook.md:
````diff
@@ -0,0 +1,271 @@
+The tool respects exclusion rules defined in:
+
+```
+packages/contracts-bedrock/scripts/checks/test-validation/exclusions.toml
````
Hmm, why are we reusing the exclusions from here?

aliersh replied:
Since a significant part of the prompt has to do with the style rules of the tests that the excluded tests do not follow, I considered it somewhat counterproductive to work on these tests with rules that do not apply to them

smartcontracts reviewed (commented):
Left a few comments for concrete things to change.

My general feedback is that the PR feels a bit over-engineered for an initial version of the code. Everything is very modular, the files are small, the functions are small, there's lots of documentation, this all makes sense from a long-term maintenance standpoint but we're very early in the development process so it creates a lot of overhead to maintain when things are likely to change rapidly. 

The main potential downside here is that all of these extra comments/docs/files need to get rewritten over and over again because we're iterating frequently. This would make review a lot harder as opposed to simply documenting and cleaning up the structure once we're happy with the product.

I see two potential futures here:
1. Our AI usage is good enough that maintaining a higher quality bar from the beginning is easy, and we don't struggle much down the line
2. We end up doing huge rewrites that delete big functions, change the docs constantly, etc etc. in a way that makes reviewing the changes hard when things are evolving rapidly

I know the whole point of this project is to dive into more AI tooling, we should definitely be using that tooling to produce more higher-quality code from the beginning, so I'm just going to roll with it for now, but we should keep an eye on it and see which future ends up playing out. Will be a good experiment for future usage.

aliersh committed:
fix(ops): clean up AI Contract Test Maintenance System files

- Remove generated ranking.json output file
- Remove CHANGELOG.md
- Fix formatting issues in runbook.md documentation

aliersh committed:
refactor(ops): simplify AI Contract Test Maintenance System structure

- consolidate modular Python files into single test_ranker.py module
- trim runbook.md to essential info only (overview, usage, output)
- add .gitignore for Python projects and output files

aliersh committed:
feat(ai-eng): add exclusion configuration for contract test maintenance

- Create exclusion.toml file with separate directory and file exclusions
- Update test_ranker.py to use centralized exclusion configuration
- Add tests without .sol counterparts to exclusion list
- Apply code formatting improvements to test_ranker.py
